torch/cuda/_memory_viz.py

```diff
@@ -824,6 +837,586 @@ def add_element(size, tensor_key, version):
 </html>
 """
 
+def segment_plot(data: Any, device=None):
+    device = _choose_device(data, device)
+    if device is None:
+        trace = []
+    else:
+        trace = data['device_traces'][device]
+
+    string_table: List[str] = []
+    suffix_table: List[Tuple[int, Optional[int]]] = []
+
+    @cache
+    def intern_str(s):
+        string_table.append(s)
+        return len(string_table) - 1
+
+    @cache
+    def intern_suffix(sid, restid):
+        suffix_table.append((sid, restid))
+        return len(suffix_table) - 1
+
+    def intern_stack(frames):
+        next_id = None
+        for f in reversed(frames):
+            next_id = intern_suffix(intern_str(f), next_id)
+        return next_id
+
+    def format_frames(frames):
+        return intern_stack([_frame_fmt(f, full_filename=True) for f in frames])
+
+    result: Any = {
+        'string_table': string_table,
+        'suffix_table': suffix_table,
+        'events': {
+            'action': [],  # reference to string table
+            'addr': [],  # for OOM, this will hold device_free value
+            'size': [],
+            'stream': [],
+            'frames': []  # reference to suffix_table
+        },
+        'segments': {
+            'addr': [],
+            'size': [],
+            'stream': []
+        },
+        'blocks': {
+            'addr': [],
+            'size': [],
+            'real_size': [],
+            'frames': [],  # reference to string table
+            'pending_free': [],
+        }
+    }
+
+    def fold_free(ts):
+        # turn a free_requested/free_completed pair into a single free event
+        i = 0
+        while i < len(ts):
+            t = ts[i]
+            if i + 1 < len(ts):
+                tnext = ts[i + 1]
+                if t['action'] == 'free_requested' and tnext['action'] == 'free_completed' and t['addr'] == tnext['addr']:
```

Comment:
requested/completed pairs on the same address are guaranteed to be next to each other I guess?

Reply:
They aren't always (sometimes the completed gets deferred), but that is the rare case, so it is nice to show most things as a single free event, and then highlight the rare cases by splitting the requested/completed events.